### PR TITLE
Add get step function data

### PR DESCRIPTION
**Pull request by halink0803**
TODO tasks:

- [ ] ~Wrap Blockchain ContractCaller to match abi/bind.ContractCaller interface~
- [x] Concurrent getStepFunction
- [x] Prefetch step function data

I change to use our Contract.Call function, so don't need to wrap our blockchain contract caller anymore.

### CODE REVIEW

**Comment by halink0803** [comment]
TODO:
- [ ] Check valid data from blockchain

**Comment by halink0803** [comment]
[Create an endpoint which return all imbalanceStepFunc of supported token](https://trello.com/c/TGXqBomo/234-create-an-endpoint-which-return-all-imbalancestepfunc-of-supported-token)

File: data/fetcher/fetcher.go [hunk 1]
```diff
@@ -743,3 +747,87 @@ func (self *Fetcher) fetchPriceFromExchange(wg *sync.WaitGroup, exchange Exchang
...
+			data.Range(func(k, v interface{}) bool {
+				token := k.(string)
+				stepData := v.(common.StepFunctionResponse)
+				result.Tokens[token] = stepData
+				return true
+			})
+
+			// save data to storage
+			err := fc.stepFunctionDataStorage.StoreStepFunctionData(result)
+			if err != nil {
```
Comment (favadi) [inline data/fetcher/fetcher.go]: using same error instance as outer scope.

File: data/fetcher/fetcher.go [hunk 2]
```diff
@@ -743,3 +747,87 @@ func (self *Fetcher) fetchPriceFromExchange(wg *sync.WaitGroup, exchange Exchang
...
+	}
+}
+
+func (fc *Fetcher) fetchStepFunctionData(data *sync.Map, block uint64) error {
+	var fetchStepFunctionDataErrCh = make(chan error)
+
+	// get internal token list
+	tokens, err := fc.setting.GetInternalTokens()
+	if err != nil {
+		log.Printf("Cannot get internal tokens: %s", err.Error())
```
Comment (favadi) [inline data/fetcher/fetcher.go]: this should be a fatal error.

File: data/storage/bolt.go [hunk 3]
```diff
@@ -1670,3 +1674,41 @@ func (self *BoltStorage) ConfirmTokenUpdateInfo(tarQty common.TokenTargetQtyV2,
...
+	err := bs.db.Update(func(tx *bolt.Tx) error {
+		dataJSON, uErr := json.Marshal(data)
+		if uErr != nil {
+			return uErr
+		}
+		b := tx.Bucket([]byte(stepFunctionBucket))
+		v := b.Get(key)
+		if v != nil {
+			uErr := b.Delete(key)
+			if uErr != nil {
```
Comment (favadi) [inline data/storage/bolt.go]: I don't think it is necessary to delete the data before put the new key.

File: data/storage/bolt.go [hunk 4]
```diff
@@ -1670,3 +1674,41 @@ func (self *BoltStorage) ConfirmTokenUpdateInfo(tarQty common.TokenTargetQtyV2,
...
+			}
+		}
+		return b.Put(key, dataJSON)
+	})
+	return err
+}
+
+//GetStepFunctionData return current step function data
+func (bs *BoltStorage) GetStepFunctionData() (common.StepFunctionData, error) {
+	key := []byte("latest_data")
```
Comment (favadi) [inline data/storage/bolt.go]: this key should be a package level const.

**Comment by favadi** [comment]
I pushed some minor fixes in https://github.com/KyberNetwork/reserve-data/pull/420/commits/3c2e67b52285d5d3ffb66d0c32bd82a67174a778.

**Pull request merged**